src/_pytest/assertion/rewrite.py

```diff
@@ -276,33 +252,32 @@ def load_module(self, name):
             # the reload() builtin will not work correctly.)
             mod = sys.modules[name]
         else:
-            # I wish I could just call imp.load_compiled here, but __file__ has to
-            # be set properly. In Python 3.2+, this all would be handled correctly
-            # by load_compiled.
-            mod = sys.modules[name] = imp.new_module(name)
+            spec = spec_from_file_location(name, co.co_filename, loader=self)
+            mod = sys.modules[name] = importlib.util.module_from_spec(spec)
         try:
-            mod.__file__ = co.co_filename
-            # Normally, this attribute is 3.2+.
-            mod.__cached__ = pyc
-            mod.__loader__ = self
-            # Normally, this attribute is 3.4+
-            mod.__spec__ = spec_from_file_location(name, co.co_filename, loader=self)
             exec(co, mod.__dict__)
-        except:  # noqa
+        except BaseException:
             if name in sys.modules:
                 del sys.modules[name]
             raise
         return sys.modules[name]
 
+    @classmethod
+    def _find_dotted_spec(cls, name):
+        pkg, _, rest = name.rpartition(".")
+        if pkg:
+            base = cls._find_dotted_spec(pkg)
+            if base is None:
+                return None
+            else:
+                return cls._find_spec(rest, path=base.submodule_search_locations)
+        else:
+            return cls._find_spec(name)
+
     def is_package(self, name):
-        try:
-            fd, fn, desc = self._imp_find_module(name)
-        except ImportError:
-            return False
-        if fd is not None:
-            fd.close()
-        tp = desc[2]
-        return tp == imp.PKG_DIRECTORY
+        _, _, modname = name.rpartition(".")
+        spec = self._find_dotted_spec(name)
+        return spec and spec.loader.is_package(modname)
```

Comment:
this is by far my least favorite part of this patch :S -- I couldn't find a better way to do this though